### PR TITLE
Audio: basefw: Implement ipc4 modules info get

### PR DESCRIPTION
**Pull request by gbernatxintel**
Add support to handle IPC4_MODULES_INFO_GET message.
This make it possible to get information about the current loaded modules.

### CODE REVIEW

File: src/audio/base_fw.c [hunk 1]
```diff
@@ -497,6 +520,7 @@ static int basefw_get_large_config(struct comp_dev *dev,
 	case IPC4_DSP_RESOURCE_STATE:
 	case IPC4_NOTIFICATION_MASK:
 	case IPC4_MODULES_INFO_GET:
+		return  basefw_modules_info_get(data_offset, data);
```
Comment (lyakh) [inline src/audio/base_fw.c:523]: this handler will also handle `IPC4_DSP_RESOURCE_STATE` and `IPC4_NOTIFICATION_MASK`?

File: src/audio/base_fw.c [hunk 2]
```diff
@@ -419,6 +421,27 @@ static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
...
+		return -EINVAL;
+	}
+
+	module_info->modules_count = desc->header.num_module_entries;
+
+	for(int idx = 0; idx < module_info->modules_count;  ++idx){
+		struct sof_man_module *module_entry =
+			(struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(idx));
+		memcpy(&(module_info->modules[idx]), module_entry, sizeof(struct sof_man_module));
+		}
```
Comment (lyakh) [inline src/audio/base_fw.c:439]: indentation

File: src/audio/base_fw.c [hunk 3]
```diff
@@ -419,6 +421,27 @@ static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
...
+	struct ipc4_modules_info *const module_info = (struct ipc4_modules_info *)data;
+
+	struct sof_man_fw_desc *desc = (struct sof_man_fw_desc *)IMR_BOOT_LDR_MANIFEST_BASE;
+	if (!desc) {
+		return -EINVAL;
+	}
+
+	module_info->modules_count = desc->header.num_module_entries;
+
+	for(int idx = 0; idx < module_info->modules_count;  ++idx){
```
Comment (lyakh) [inline src/audio/base_fw.c:435]: space before `{`

File: src/audio/base_fw.c [hunk 4]
```diff
@@ -419,6 +421,27 @@ static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
...
+		struct sof_man_module *module_entry =
+			(struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(idx));
+		memcpy(&(module_info->modules[idx]), module_entry, sizeof(struct sof_man_module));
```
Comment (lyakh) [inline src/audio/base_fw.c:438]: `sizeof(module_info->modules[idx])` as the last parameter

File: src/audio/base_fw.c [hunk 5]
```diff
@@ -419,6 +421,27 @@ static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
...
+		}
+
+	*data_offset = sizeof(uint32_t) + module_info->modules_count * sizeof(struct sof_man_module);
```
Comment (lyakh) [inline src/audio/base_fw.c:441]: I'd do `sizeof(module_info->modules_count) + module_info->modules_count * sizeof(module_info->modules[0])` - makes it clearer what is calculated IMHO

**Review by lyakh — CHANGES_REQUESTED** [review]

File: src/audio/base_fw.c [hunk 6]
```diff
@@ -419,6 +421,27 @@ static int basefw_power_state_info_get(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static int basefw_modules_info_get(uint32_t * data_offset, char* data)
+{
+	struct ipc4_modules_info *const module_info = (struct ipc4_modules_info *)data;
+
+	struct sof_man_fw_desc *desc = (struct sof_man_fw_desc *)IMR_BOOT_LDR_MANIFEST_BASE;
```
Comment (kv2019i) [inline src/audio/base_fw.c:428]: This is a bit platform specific. The work to add a platform layer for IPC4 basefw is still in progress, but at a minimum, I think you need to ifdef this with "#ifdef RIMAGE_MANIFEST" like done in ipc4/helper.c.

File: src/include/ipc/header.h [hunk 7]
```diff
@@ -329,7 +329,7 @@ struct ipc_cmd_hdr;
 #define SOF_IPC_MESSAGE_ID(x)			((x) & 0xffff)
 
 /** Maximum message size for mailbox Tx/Rx */
-#define SOF_IPC_MSG_MAX_SIZE			384
```
Comment (kv2019i) [inline src/include/ipc/header.h:332]: I think this needs a separate commit and rationale for the increase. This affects all IPC3 implementations as well and increases memory use as this is directly used to allocate memory (which so far has been 384 bytes). 

Or is this something only for IPC4. Then tihs could be changed only if built for IPC4.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Some review comments inline.

File: src/include/ipc/header.h (hunk 7 above)
Comment (lgirdwood) [reply src/include/ipc/header.h:332]: @gbernatxintel We can play safe here. e.g.
```
#if IPC4
#define SIZE 0x1000
#else
#define SIZE 384
#endif
```

Comment (abonislawski) [reply src/include/ipc/header.h:332]: IPC fuzz failed only for IPC3 so please try this @gbernatxintel

Comment (gbernatxintel) [reply src/include/ipc/header.h:332]: Done

File: src/include/ipc4/base_fw.h [hunk 8]
```diff
@@ -17,6 +17,10 @@
 #ifndef __SOF_IPC4_BASE_FW_H__
 #define __SOF_IPC4_BASE_FW_H__
 
+#ifdef CONFIG_INTEL
```
Comment (lgirdwood) [inline src/include/ipc4/base_fw.h:20]: Can we make this a new Kconfig as manifest data is useful and generic that could be used by anyone. e.g. ```KCONFIG_IPC4_MANIFEST_MODULES``` we can have other ipc4 manifest kconfisg too as I would imagine a lot of vendors would like to export data.

File: src/include/ipc/header.h [hunk 9]
```diff
@@ -329,8 +329,11 @@ struct ipc_cmd_hdr;
 #define SOF_IPC_MESSAGE_ID(x)			((x) & 0xffff)
 
 /** Maximum message size for mailbox Tx/Rx */
+#if IPC4
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/include/ipc/header.h:332]: @gbernatxintel use `#ifdef CONFIG_IPC_MAJOR_4`

**Comment by gbernatxintel** [comment]
Update
Moved module_info_get functions to platform specs file, as there were problems with ipc fuzz testing. In general, the problem was that I use a structure that is defined in rimage/manifest.h. This file occurs on the intel platform, so to avoid the use of ifdef the whole thing was moved to the platform file.

@marc-hb Could you take a look at CI, is there a compilation problem?

File: src/include/ipc/header.h (hunk 7 above)
Comment (kv2019i) [reply src/include/ipc/header.h:332]: @ujfalusi @plbossart This SOF_IPC_MSG_MAX_SIZE is only used in IPC3 code, so if this change is ifdeffed like this, I don't think we need to bump the ABI ver

File: src/include/kernel/abi.h [hunk 10]
```diff
@@ -29,7 +29,7 @@
 
 /** \brief SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 29
+#define SOF_ABI_MINOR 30
```
Comment (kv2019i) [inline src/include/kernel/abi.h:32]: So this is not needed as this ABI versioning applies to IPC3 builds.

We still don't have infra to describe IPC4 infra. As this has been pending for long, let me file a ticket -> https://github.com/thesofproject/sof/issues/9080

**Review by kv2019i — COMMENTED** [review]
#9060 was merged, so you'll need to rebase this.

File: src/audio/base_fw.c [hunk 11]
```diff
@@ -461,7 +466,9 @@ static int basefw_get_large_config(struct comp_dev *dev,
 	/* TODO: add more support */
 	case IPC4_DSP_RESOURCE_STATE:
 	case IPC4_NOTIFICATION_MASK:
+		break;
```
Comment (marc-hb) [inline src/audio/base_fw.c:469]: Why is this not even printing a warning?

Comment (marc-hb) [reply src/audio/base_fw.c:469]: You seem to be splitting the "TODO", unimplemented section in the middle. This means the second part is losing its "TODO" comment and silently doing nothing with zero comment and zero warning. So now it looks even less intentional and even more like a bug.

File: src/include/ipc/header.h (hunk 7 above)
Comment (ujfalusi) [reply src/include/ipc/header.h:332]: @kv2019i, the kernel uses different defines for the message size (in mailbox):
include/sound/sof/header.h:#define SOF_IPC_MSG_MAX_SIZE                 384
include/sound/sof/ipc4/header.h:#define SOF_IPC4_MSG_MAX_SIZE                   4096

sound/soc/sof/ipc3.c:   sdev->ipc->max_payload_size = SOF_IPC_MSG_MAX_SIZE;
sound/soc/sof/ipc4.c:   sdev->ipc->max_payload_size = SOF_IPC4_MSG_MAX_SIZE;

File: src/audio/base_fw.c (hunk 11 above)
Comment (ujfalusi) [reply src/audio/base_fw.c:469]: move out the IPC4_MODULES_INFO_GET case from the TODO list.

File: src/audio/base_fw.c [hunk 12]
```diff
@@ -461,7 +466,9 @@ static int basefw_get_large_config(struct comp_dev *dev,
...
 	case IPC4_MODULES_INFO_GET:
```
Comment (ujfalusi) [inline src/audio/base_fw.c:470]: The commit message can use some update.
"Added support for ipc4 query no 9." 
->
"Add support to handle IPC4_MODULES_INFO_GET message."

File: src/audio/base_fw.c [hunk 13]
```diff
@@ -334,6 +334,11 @@ static int basefw_libraries_info_get(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static int basefw_modules_info_get(uint32_t *data_offset, char *data)
+{
+	return platform_basefw_modules_info_get(data_offset, data);
```
Comment (ujfalusi) [inline src/audio/base_fw.c:339]: why it is offloaded to platform function?
What is the difference between the module info and library info (`basefw_libraries_info_get()`) to demand this handling?

File: src/audio/base_fw.c (hunk 11 above)
Comment (gbernatxintel) [reply src/audio/base_fw.c]: done, currently all implemented cases are before the TODO comment

File: src/audio/base_fw.c (hunk 12 above)
Comment (gbernatxintel) [reply src/audio/base_fw.c]: Fixed

File: src/audio/base_fw.c (hunk 13 above)
Comment (gbernatxintel) [reply src/audio/base_fw.c:339]: I have generally separated the function platform_basefw_modules_info_get(...) instead of implementing everything in base_fw.c because I use the ipc4_modules_info structure which uses sof_man_module.
sof_man_module is defined in rimage/manifest.h

I had a problem with the tests because on non-Intel platforms there is no manifest.h file and the tests did not pass. Besides, this modular approach follows the convention that we create code in base_fw_platform.c

With regard to the differences between basefw_modules_info_get and basefw_libraries_info_get

This basefw_libraries_info_get returns the currently loaded library and other structure. 
 basefw_modules_info_get returns which modules are loaded and which are not, along with information about them. 

In short, both commands return a different data type, different structure. In preparing this PR I looked at previous code and tests that need specific information.

**Review by lgirdwood — COMMENTED** [review]
@kv2019i @lyakh good for you now ?

**Review by kv2019i — APPROVED** [review]

File: src/platform/intel/ace/lib/base_fw_platform.c [hunk 14]
```diff
@@ -72,6 +78,28 @@ struct sof_man_fw_desc *platform_base_fw_get_manifest(void)
...
+	module_info->modules_count = desc->header.num_module_entries;
+
+	for (int idx = 0; idx < module_info->modules_count; ++idx) {
+		struct sof_man_module *module_entry =
+			(struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(idx));
+		memcpy_s(&module_info->modules[idx], sizeof(module_info->modules[idx]),
+			 module_entry, sizeof(struct sof_man_module));
+	}
+
+	*data_offset = sizeof(module_info) +
```
Comment (lyakh) [inline src/platform/intel/ace/lib/base_fw_platform.c:98]: you meant `sizeof(*module_info)`

**Review by lyakh — COMMENTED** [review]
one error to correct

File: src/platform/intel/ace/lib/base_fw_platform.c (hunk 14 above)
Comment (gbernatxintel) [reply src/platform/intel/ace/lib/base_fw_platform.c]: Good point, fixed

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@lyakh good now ?

**Review by lyakh — APPROVED** [review]
